### PR TITLE
Make sure that remote shares use the correct uid casing

### LINKED ISSUE #19399: Federated cloud sharing requires case sensitive user names.
### Steps to reproduce
1. Share a file with an all-lowercase user on a different owncloud-instance using a mixed-case username in sharing dialog.

This is more obviously when the user in question actually uses a mixed-case username and sharing is performed using all-lowercase.

### Expected behaviour
Regardless of case of characters, the shared file should show up in the target user's files app. At least after the target user changed apps for once.

### Actual behaviour
When the case of the username in the sharing dialog does not match the actual username's case, the share does not show up. It does so only after the username is written exactly as the registered username.

ownCloud is case-insensitive on logins - and I haven't tried whether or not, but doubt it, probably does not accept a username which matches another one except case, yet when using federated cloud sharing, usernames are case-sensitive.

The behaviour should at least be consistent, preferrably being case-insensitive on sharing.

### Server configuration
**Operating system:** Linux, 4.1 Kernel series.

**Web server:** Apache-2.4 on source instance, Nginx-1.8.0 on target instance.

**Database:** MariaDB-10.0.21

**PHP version:** 5.6.13

**ownCloud version:** 8.1.3

**Updated from an older ownCloud or fresh install:** Both updated, source instance has the full history from 4.5.7 through 8.1.3, target instance has started with 8.x series.

**Are you using external storage, if yes which one:** no

**Are you using encryption:** Source instance - no, target instance - yes.

**Are you using an external user-backend, if yes which one:** no

### Client configuration
**Browser:** Seamonkey-2.3x as the source sharer, Firefox-38.x as the share recipient.

**Operating system:** Both on the same Linux system with 4.0 kernel series.

### PR DESCRIPTION
**Pull request by nickvergessen**
Fixes #19399 

@schiesbn @PVince81 @ftiede 

PS: Only works for new shares

### CODE REVIEW

**Comment by karlitschek** [comment]
:+1:

**Comment by schiesbn** [comment]
looks good :+1:

**Comment by icewind1991** [comment]
I would prefer to explicitly get the correct uid in the s2s api instead.

**Comment by icewind1991** [comment]
Instead of adding code to core which should be in the share app instead

**Comment by nickvergessen** [comment]
That is exactly what it does... just via an existing hook ldap already uses to translate names.

File: lib/private/user/database.php [hunk 1]
```diff
@@ -302,4 +316,22 @@ public function getBackendName(){
 		return 'Database';
 	}
 
+	public static function preLoginNameUsedAsUserName($param) {
+		if(!isset($param['uid'])) {
+			throw new \Exception('key uid is expected to be set in $param');
+		}
+
+		$backends = \OC::$server->getUserManager()->getBackends();
+		foreach ($backends as $backend) {
```
Comment (icewind1991) [inline lib/private/user/database.php]: I mean moving this logic to the s2s api

File: lib/base.php [hunk 2]
```diff
@@ -606,6 +606,14 @@ public static function init() {
 		OC_User::useBackend(new OC_User_Database());
 		OC_Group::useBackend(new OC_Group_Database());
 
+		// Subscribe to the hook
+		\OCP\Util::connectHook(
```
Comment (icewind1991) [inline lib/base.php]: Or at the minimum not having core handle app hooks

File: lib/private/user/database.php (hunk 1 above)
Comment (icewind1991) [inline lib/private/user/database.php]: Even if this was to remain in core it shouldn't be inside a backend

File: lib/base.php (hunk 2 above)
Comment (nickvergessen) [inline lib/base.php]: Added a FIXME to the place triggering the hook, that it should actually be a method in the user management.
Don't want to be too explosive right before a release

**Comment by nickvergessen** [comment]
@MorrisJobke review? ;)

**Comment by MorrisJobke** [comment]
Tested and works :+1:

**Comment by DeepDiver1975** [comment]
:-1: for this - https://github.com/owncloud/core/issues/19399 has no milestone and no severity

**Comment by MorrisJobke** [comment]
> :-1: for this - #19399 has no milestone and no severity

Then simply move the milestone and keep it open until the stable8.2 is branched off

**Comment by nickvergessen** [comment]
@DeepDiver1975 whats the procedure? Ask cml if we can get this for 8.2.1, or is it 9.0 and then ask for a backport, or ask cml for 8.2.0 permissions?

**Comment by nickvergessen** [comment]
@karlitschek should backport this to 8.2.1

**Comment by karlitschek** [comment]
please backport

**Pull request merged**